metadata/src/main/java/org/apache/kafka/controller/ActivationRecordsGenerator.java

```diff
@@ -0,0 +1,197 @@
+            bootstrapMetadata.records().size(),
+            metadataVersion,
+            bootstrapMetadata.source());
+        records.addAll(bootstrapMetadata.records());
+
+        if (metadataVersion.isMigrationSupported()) {
+            if (zkMigrationEnabled) {
+                log.info("Putting the controller into pre-migration mode. No metadata updates will be allowed " +
+                    "until the ZK metadata has been migrated");
+                records.add(ZkMigrationState.PRE_MIGRATION.toRecord());
+            } else {
+                log.debug("Setting the ZK migration state to NONE since this is a de-novo KRaft cluster.");
+                records.add(ZkMigrationState.NONE.toRecord());
+            }
+        } else {
+            if (zkMigrationEnabled) {
+                throw new RuntimeException("The bootstrap metadata.version " + bootstrapMetadata.metadataVersion() +
+                    " does not support ZK migrations. Cannot continue with ZK migrations enabled.");
+            }
+        }
+
+        if (metadataVersion.isMetadataTransactionSupported()) {
+            records.add(new ApiMessageAndVersion(new EndTransactionRecord(), (short) 0));
+            return ControllerResult.of(records, null);
+        } else {
+            return ControllerResult.atomicOf(records, null);
+        }
+    }
+
+    static ControllerResult<Void> recordsForNonEmptyLog(
+        Logger log,
+        long transactionStartOffset,
+        boolean zkMigrationEnabled,
+        FeatureControlManager featureControl,
+        MetadataVersion metadataVersion
+    ) {
+        // Logs have been replayed. We need to initialize some things here if upgrading from older KRaft versions
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+
+        if (metadataVersion.equals(MetadataVersion.MINIMUM_KRAFT_VERSION)) {
+            log.info("No metadata.version feature level record was found in the log. " +
+                "Treating the log as version {}.", MetadataVersion.MINIMUM_KRAFT_VERSION);
+        }
+
+        if (metadataVersion.isMigrationSupported()) {
+            log.info("Loaded ZK migration state of {}", featureControl.zkMigrationState());
+            switch (featureControl.zkMigrationState()) {
+                case NONE:
+                    // Since this is the default state there may or may not be an actual NONE in the log. Regardless,
+                    // it will eventually be persisted in a snapshot, so we don't need to explicitly write it here.
+                    if (zkMigrationEnabled) {
+                        throw new RuntimeException("Should not have ZK migrations enabled on a cluster that was " +
+                            "created in KRaft mode.");
+                    }
+                    break;
+                case PRE_MIGRATION:
+                    log.warn("Activating pre-migration controller without empty log. There may be a partial migration");
```

Comment:
I think if the MV doesn't support transactions, we should throw here

If the MV does support transactions, we shouldn't scare people about "partial migrations" since that shouldn't be happening, but just state that we haven't yet loaded the initial ZK state

Reply:
Can we throw here? If a controller in migration mode is activated it will write a PRE_MIGRATION record. Since it may take a while for the migration to begin (e.g., brokers aren't all configured for migration yet), we could have a controller failover. In that case, we wouldn't have started the migration and would still be in PRE_MIGRATION mode.

The "There may be a partial migration" comes because there may in fact be a partial migration if the controller failed over before the MIGRATION record was written. 

If we throw here, then we can't tolerate controller failover while waiting for the migration to kick off.

Reply:
The proposal was:

Transactions enabled => we tolerate failover during PREMIGRATION, no scary log message
Transactions not enabled => we throw here, and the cluster requires manual recovery

Reply:
> Transactions enabled => we tolerate failover during PREMIGRATION, no scary log message

This is done now in the current PR.

> Transactions not enabled => we throw here, and the cluster requires manual recovery

As discussed offline, we'll do this one as a follow-up [KAFKA-15381](https://issues.apache.org/jira/browse/KAFKA-15381)